providers/implementations/encode_decode/decode_der2key.c

```diff
@@ -552,6 +575,234 @@ static void *sm2_d2i_PKCS8(const unsigned char **der, long der_len,
+ * -  2 byte algorithm sequence tag and length
+ * -    2 byte algorithm OID tag and length
+ * -      9 byte algorithm OID (from NIST CSOR OID arc)
+ * -  4 byte bit string tag and length
+ * -    1 bitstring lead byte
+ */
+#define ML_KEM_SPKI_OVERHEAD   22
+
+/*
+ * The accepted private key format is either the 64-bytes (d, z) seed pair, or
+ * else the optionally DER wrapped octet-string encoding  (one byte tag, three
+ * byte length) of the FIPS 203 expanded |dk| key, optionally concatenated with
+ * the public |ek|.  We don't need to worry about whether the public key is or
+ * isn't appended, those are just ignored bytes at the end of an input buffer.
+ */
+typedef enum {
+    NONE_FMT,   /* Unknown or unsupported format */
+    SEED_FMT,   /* Just the seed, no DER wrapping */
+    LONG_FMT,   /* Key(s), no octet-string DER wrapping */
+    ASN1_FMT    /* Key(s), octet-string DER wrapped */
+} ML_KEM_PRIV_FMT;
+
+static void *
+ml_kem_d2i_PKCS8(const uint8_t **der, long der_len, struct der2key_ctx_st *ctx)
+{
+    const ML_KEM_VINFO *vinfo;
+    ML_KEM_KEY *key = NULL, *ret = NULL;
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(ctx->provctx);
+    PKCS8_PRIV_KEY_INFO *p8inf = NULL;
+    const unsigned char *p;
+    const X509_ALGOR *alg = NULL;
+    ML_KEM_PRIV_FMT fmt;
+    int plen, ptype, privlen, pairlen;
+    uint16_t keylen;
+
+    /* Extract the key OID and any parameters (we want none of those) */
+    if ((p8inf = d2i_PKCS8_PRIV_KEY_INFO(NULL, der, der_len)) == NULL)
+        return 0;
+    if (!PKCS8_pkey_get0(NULL, &p, &plen, &alg, p8inf))
+        goto end;
+
+    /* Bail out early if this is some other key type. */
+    if (OBJ_obj2nid(alg->algorithm) != ctx->desc->evp_type)
+        goto end;
+    vinfo = ossl_ml_kem_get_vinfo(ctx->desc->evp_type);
+
+    /* Parameters must be absent. */
+    X509_ALGOR_get0(NULL, &ptype, NULL, alg);
+    if (ptype != V_ASN1_UNDEF) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_UNEXPECTED_KEY_PARAMETERS,
+                       "unexpected parameters with a PKCS#8 %s private key",
+                       vinfo->algorithm_name);
+        goto end;
+    }
+    privlen = (int) vinfo->prvkey_bytes;
+    pairlen = (int) (vinfo->prvkey_bytes + vinfo->pubkey_bytes);
+
+    if (plen == ML_KEM_SEED_BYTES)
+        fmt = SEED_FMT;
+    else if (plen == privlen || plen == pairlen)
+        fmt = LONG_FMT;
+    else if (plen == ML_KEM_OCTET_STRING_OVERHEAD + privlen
+             || plen == ML_KEM_OCTET_STRING_OVERHEAD + pairlen)
+        fmt = ASN1_FMT;
+    else
+        fmt = NONE_FMT;
+
+    switch (fmt) {
+    case NONE_FMT:
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH,
+                       "unexpected PKCS#8 private key length: %d", plen);
+        break;
+    case SEED_FMT:
+        if ((key = ossl_ml_kem_key_new(libctx, ctx->propq,
+                                       ctx->desc->evp_type)) != NULL
+            && ossl_ml_kem_genkey(p, p + ML_KEM_RANDOM_BYTES, NULL, 0, key))
```

Comment:
I assume the genkey issue in the decoder and its impact on FIPS will be sorted out in a subsequent PR

Reply:
We're OK for the moment because the FIPS provider regenerates from the seed, which still happens to be be the export format for now. Though, we did agree to change the **export** and **import** formats to carry an explicit **seed** parameter with the **priv** element back to |dk|, which does not materially affect the encoder/decoder, because it does not know or care about what export/import do, so long as they magically work.

Changing the **export** / **import** param sets is on the short-term TODO list. But not related to encoding decoding in this PR.

Reply:
Thanks for the review. Much appreciated.

Reply:
> Though, we did agree to change the export and import formats to carry an explicit seed parameter with the priv element back to |dk|, which does not materially affect the encoder/decoder, because it does not know or care about what export/import do, so long as they magically work.

Which matches with what I was saying in my other comment

Reply:
Sorry, my confusion, we're in violent agreement. Thanks!